examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -41,101 +50,346 @@ def freq_mask(freqs, fmin, fmax):
+
+        Parameters
+        ----------
+        inst : instance of Raw or Epochs (n_epochs, n_channels, n_times)
+            The data to be processed. The instance is modified inplace.
+       
+        Returns
+        -------
+        out : instance of Raw or Epochs
+            The processed data.
+        
+        """
+        if self.filters_ is None:
+            raise RuntimeError('No filters available. Please first call fit')
+        if self.return_filtered:
+            inst=self.Xs
+        if isinstance(inst, BaseRaw):
+            data=inst.get_data()
+            X_ssd=np.dot(self.filters_.T, data[self.picks_])
+           
+        else:
+            if isinstance(inst, BaseEpochs)  or isinstance(inst, np.ndarray):
+                
+                data=inst    
+                #project data on source space
+                X_ssd = np.asarray([np.dot(self.filters_.T, epoch) for epoch in data])
+        
+        if self.sort_by_spectral_ratio:
+            self.spec_ratio, self.sorter_spec=self.spectral_ratio_ssd(ssd_sources=X_ssd)
+            self.filters_=self.filters_[:,self.sorter_spec]
+            self.patterns_=self.patterns_[self.sorter_spec]
+
+            if isinstance(inst, BaseRaw):
+                X_ssd=X_ssd[self.sorter_spec]
+            else:
+                X_ssd=X_ssd[:,self.sorter_spec,:]
+            if self.n_components is None:
+                n_components = self.max_components
+                return X_ssd
+            else:
+                n_components = self.n_components
+                if isinstance(inst, BaseRaw):
+                    X_ssd=X_ssd[:n_components]
+                else:
+                    X_ssd=X_ssd[:,:n_components,:]
+       
+    
+    def apply(self, inst):
+        """
+        Remove selected components from the signal.
+        This procedure will reconstruct M/EEG signals from which the dynamics 
+        described by the excluded components is subtracted (denoised by low-rank factorization). 
+        See [2]  Haufe et al. for more information.
+        
+        The data is processed in place.
+
+        Parameters
+        ----------
+        inst : instance of Raw or Epochs 
+            The data to be processed. The instance is modified inplace.
+       
+       
+        Returns
+        -------
+        out : instance of Raw or Epochs
+            The processed data.
+        
+        """
+        X_ssd=self.transform(inst)
+       
+        
+        pick_patterns = self.patterns_[:self.n_components].T
+        if isinstance(inst, BaseRaw):
+            X=np.dot(pick_patterns, X_ssd)
 
-    def apply(self, instance):
-        pass
+        else:
+            if isinstance(inst, BaseEpochs):
+                if not isinstance(inst, np.ndarray):
+                    raise ValueError("X should be of type ndarray (got %s)." % type(inst))
+                X=np.asarray([np.dot(pick_patterns, epoch) for epoch in X_ssd])
```

Comment:
Probably it would be better to pre-allocate the output array and add the results step by step.